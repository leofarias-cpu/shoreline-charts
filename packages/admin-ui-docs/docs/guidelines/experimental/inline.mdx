---
title: Inline
path: /inline/
---

Render a set of components in a row with equal spacing between them. It also supports wrapping.

## Usage

```jsx
import { experimental_Inline as Inline } from '@vtex/admin-ui'

function Example() {
  return (
    <Inline>
      <Tag label="Brazil" />
      <Tag label="Argentina" />
      <Tag label="Chile" />
      <Tag label="Uruguai" />
    </Inline>
  )
}
```

## Examples

### Basic

```jsx noInline live
const Inline = experimental_Inline

function Example() {
  return (
    <Inline>
      <Tag label="Brazil" />
      <Tag label="Argentina" />
      <Tag label="Chile" />
      <Tag label="Uruguai" />
    </Inline>
  )
}

render(<Example />)
```

### Wrap

Wrapping is applied automatically.

```jsx noInline live
const Inline = experimental_Inline

function Example() {
  return (
    <Box csx={{ width: '18rem' }}>
      <Inline>
        <Tag label="Brasil" />
        <Tag label="Argentina" />
        <Tag label="Chile" />
        <Tag label="Uruguai" />
        <Tag label="Colombia" />
        <Tag label="Equador" />
        <Tag label="Venezuela" />
      </Inline>
    </Box>
  )
}

render(<Example />)
```

### Space

Space is configured using the `vSpace` and `hSpace`, with either token or number values. In the case of a numeric value, the unit is `rem` by default. You can also set the `unit` prop to change it.

```jsx noInline live
const Inline = experimental_Inline

function Example() {
  return (
    <Box csx={{ width: '18rem' }}>
      <Inline vSpace="$xl" hSpace="$xl">
        <Tag label="Brasil" />
        <Tag label="Argentina" />
        <Tag label="Chile" />
        <Tag label="Uruguai" />
        <Tag label="Colombia" />
        <Tag label="Equador" />
        <Tag label="Venezuela" />
      </Inline>
    </Box>
  )
}

render(<Example />)
```

## Props

| Name   | Type                     | Description                         | Required | Default |
| ------ | ------------------------ | ----------------------------------- | -------- | ------- |
| vSpace | `VSpaceTokens or number` | Vertical space                      | ðŸš«       | '$s'    |
| hSpace | `HSpaceTokens or number` | Horizontal space                    | ðŸš«       | '$s'    |
| unit   | `CSSUnit`                | Unit used in case of a number value | ðŸš«       | 'rem'   |

