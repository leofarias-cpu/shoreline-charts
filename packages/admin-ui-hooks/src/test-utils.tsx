import 'jest-location-mock'

export function jestWindowMock() {
  let history: string[] = [window.location.pathname]
  let index = 0
  let popstateCallback = () => {}

  Object.defineProperties(window.history, {
    pushState: {
      value: (data: { path: string }) => {
        history = [...history.splice(0, index + 1), data.path]
        index += 1

        window.location.assign(data.path)
        popstateCallback()
      },
    },
    back: {
      value: () => {
        if (index <= 0) return

        index -= 1
        window.location.assign(history[index])
        popstateCallback()
      },
    },
    forward: {
      value: () => {
        if (index >= history.length) return

        index += 1
        window.location.assign(history[index])
        popstateCallback()
      },
    },
  })
  Object.defineProperties(window, {
    addEventListener: {
      value: (name: string, fn: () => void) => {
        if (name === 'popstate') {
          popstateCallback = fn
        }
      },
    },
  })
}

export function setQueryParam(query: Record<string, string | undefined>) {
  const queryParams = new URLSearchParams(window.location.search)

  Object.entries(query).forEach((element: [string, string | undefined]) => {
    if (!element[1]) {
      queryParams.delete(element[0])

      return
    }

    queryParams.set(element[0], element[1])
  })

  window.location.assign(`?${queryParams.toString()}`)
}

export function getCurrentQuery() {
  const queryParams = new URLSearchParams(window.location.search)

  return Object.fromEntries(queryParams.entries())
}

export function cleanQueryParam() {
  window.location.assign('')
}
