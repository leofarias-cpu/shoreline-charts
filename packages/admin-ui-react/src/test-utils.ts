import type { ReactElement } from 'react'
import type { RenderOptions } from '@testing-library/react'
import { render as baseRender } from '@testing-library/react'
import { ThemeProvider } from './system'

function render(ui: ReactElement, options?: Omit<RenderOptions, 'queries'>) {
  return baseRender(ui, { wrapper: ThemeProvider, ...options })
}

export { render }
