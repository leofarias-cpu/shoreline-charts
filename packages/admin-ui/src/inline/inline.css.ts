import type { SpaceTokens } from '@vtex/admin-ui-core'
import { theme, dataAttr, csx, getTokenValue } from '@vtex/admin-ui-core'
import type * as CSS from 'csstype'

function getRuntimeSpace(token: string) {
  return getTokenValue(theme, 'margin', token)
}

export const inlineStyle = (props: InlineStyleProps) =>
  ({
    '--v-space': getRuntimeSpace(props.vSpace),
    '--h-space': getRuntimeSpace(props.hSpace),
    '--align': props.align,
  } as React.CSSProperties)

export const inlineTheme = csx({
  display: 'flex',
  [dataAttr('wrap', true)]: {
    flexWrap: 'wrap',
  },
  [dataAttr('wrap', false)]: {
    flexWrap: 'nowrap',
  },
  alignItems: 'var(--align)',
  [dataAttr('space-inside', false)]: {
    '> *:not(:first-child)': {
      marginLeft: 'var(--h-space)',
      marginTop: 'var(--v-space)',
    },
    '> *:is(:first-child)': {
      marginLeft: 'var(--h-space)',
      marginTop: 'var(--v-space)',
    },
  },
  [dataAttr('space-inside', true)]: {
    '> *:not(:first-child)': {
      marginLeft: 'var(--h-space)',
      marginTop: '$space-0',
    },
    '> *:is(:first-child)': {
      marginLeft: '$space-0',
      marginTop: '$space-0',
    },
  },
})

interface InlineStyleProps {
  vSpace: SpaceTokens
  hSpace: SpaceTokens
  align: CSS.Properties['alignItems']
}
