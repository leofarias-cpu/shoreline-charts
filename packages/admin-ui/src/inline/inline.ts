import { createComponent, useElement } from '@vtex/admin-ui-react'
import { withUnit } from '@vtex/admin-ui-core'
import type { HSpaceTokens, VSpaceTokens, CSSUnit } from '@vtex/admin-ui-core'

export const Inline = createComponent<'div', InlineProps>((props) => {
  const { vSpace = '$s', hSpace = '$s', unit = 'rem', ...htmlProps } = props

  return useElement('div', {
    ...htmlProps,
    baseStyle: {
      display: 'flex',
      flexWrap: 'wrap',
      '> *': {
        marginLeft: withUnit(hSpace, unit),
        marginTop: withUnit(vSpace, unit),
      },
    },
  })
})

export interface InlineProps {
  vSpace?: VSpaceTokens | number
  hSpace?: HSpaceTokens | number
  unit?: CSSUnit
}
