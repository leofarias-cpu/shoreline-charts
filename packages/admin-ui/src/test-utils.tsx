import React, { useRef } from 'react'
import { axe } from 'jest-axe'
import type { RenderOptions } from '@testing-library/react'
import { render as baseRender } from '@testing-library/react'
import { ThemeProvider } from '@vtex/admin-core'
import type { ReactElement } from 'react'
import { renderHook, act } from '@testing-library/react-hooks'

function render(ui: ReactElement, options?: Omit<RenderOptions, 'queries'>) {
  return baseRender(ui, { wrapper: ThemeProvider, ...options })
}

/**
 * HOC to wrap a component with its state hook
 * @example
 * // Cmp has the useCmp state hook
 * const StatefulCpm = withState(Cmp, () => useCmpState({ isSelected: false }))
 */
function withState(Component: any, useStateHook: () => any) {
  function StatefulComponent(props: any) {
    // Avoids an invalid hook call
    const ref = useRef(useStateHook())

    return <Component state={ref.current} {...props} />
  }

  StatefulComponent.displayName = Component.displayName

  return StatefulComponent
}

function jestMatchMedia() {
  /**
   * ðŸš§ Workaround for window.match media
   * @see https://jestjs.io/docs/en/manual-mocks#mocking-methods-which-are-not-implemented-in-jsdom
   */
  return Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  })
}

export * from '@testing-library/react'
export { render, withState, axe, jestMatchMedia, renderHook, act }
