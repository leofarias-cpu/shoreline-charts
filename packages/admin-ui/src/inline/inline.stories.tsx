import React from 'react'

import type { Meta } from '@storybook/react'

import { Inline } from './inline'
import { Tag } from '../tag'
import { csx } from '@vtex/admin-ui-core'

export default {
  title: 'admin-ui/Inline',
} as Meta

export const Basic = () => {
  return (
    <Inline>
      <Tag label="Brazil" />
      <Tag label="Argentina" />
      <Tag label="Chile" />
      <Tag label="Uruguai" />
    </Inline>
  )
}

export const Wrap = () => {
  return (
    <div className={csx({ width: '18rem' })}>
      <Inline>
        <Tag label="Brasil" />
        <Tag label="Argentina" />
        <Tag label="Chile" />
        <Tag label="Uruguai" />
        <Tag label="Colombia" />
        <Tag label="Equador" />
        <Tag label="Venezuela" />
      </Inline>
    </div>
  )
}

export const Space = () => {
  return (
    <div className={csx({ width: '18rem' })}>
      <Inline hSpace="10px" vSpace="$space-20">
        <Tag label="Brasil" />
        <Tag label="Argentina" />
        <Tag label="Chile" />
        <Tag label="Uruguai" />
        <Tag label="Colombia" />
        <Tag label="Equador" />
        <Tag label="Venezuela" />
      </Inline>
    </div>
  )
}

export const SpaceInside = () => {
  return (
    <div className={csx({ width: '18rem' })}>
      <Inline spaceInside>
        <Tag label="Brasil" />
        <Tag label="Argentina" />
        <Tag label="Chile" />
        <Tag label="Uruguai" />
        <Tag label="Colombia" />
        <Tag label="Equador" />
        <Tag label="Venezuela" />
      </Inline>
    </div>
  )
}
