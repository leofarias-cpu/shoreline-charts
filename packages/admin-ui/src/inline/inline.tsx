import type { SpaceTokens, CSSPropAutocomplete } from '@vtex/admin-ui-core'
import type * as CSS from 'csstype'
import { inlineStyle, inlineTheme } from './inline.style'
import type { ComponentPropsWithoutRef, Ref } from 'react'
import React, { forwardRef } from 'react'
import { cx } from '@vtex/admin-ui-core'

export const Inline = forwardRef(
  (props: InlineProps, ref: Ref<HTMLDivElement>) => {
    const {
      vSpace = '$space-05',
      hSpace = '$space-1',
      noWrap = false,
      align = 'start',
      spaceInside = false,
      className = '',
      ...htmlProps
    } = props

    const responsiveCssProps = inlineStyle({ vSpace, hSpace, align })

    return (
      <div
        style={responsiveCssProps}
        className={cx(inlineTheme, className)}
        data-wrap={!noWrap}
        data-space-inside={spaceInside}
        ref={ref}
        {...htmlProps}
      />
    )
  }
)

Inline.displayName = 'Inline'

export interface InlineProps extends ComponentPropsWithoutRef<'div'> {
  /**
   * Vertical space
   * @default '$space-05'
   */
  vSpace?: CSSPropAutocomplete<SpaceTokens>
  /**
   * Horizontal space
   * @default '$space-1'
   */
  hSpace?: CSSPropAutocomplete<SpaceTokens>
  /**
   * Disable wrap
   * @default false
   */
  noWrap?: boolean
  /**
   * Items vertical alignment
   * @default 'start'
   */
  align?: CSS.Properties['alignItems']
  /**
   * Don't space the container
   * @default false
   */
  spaceInside?: boolean
}
