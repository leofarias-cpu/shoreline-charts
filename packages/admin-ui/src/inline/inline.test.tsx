import React from 'react'

import { render } from '../test-utils'
import { Inline } from './inline'

describe('Inline', () => {
  it('renders', async () => {
    const { container } = render(
      <Inline>
        <button>Button 1</button>
        <button>Button 2</button>
      </Inline>
    )

    expect(container).toBeInTheDocument()
  })
})
