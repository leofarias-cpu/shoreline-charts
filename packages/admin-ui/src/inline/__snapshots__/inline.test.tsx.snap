// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline should match snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-0 >* {
  margin-left: 0.25rem;
  margin-top: 0.125rem;
}

<div
    class="emotion-0"
  >
    <button>
      Button 1
    </button>
    <button>
      Button 2
    </button>
  </div>
</DocumentFragment>
`;
